### PR TITLE
Refactor Plot Mass

### PR DESCRIPTION
**Pull request by LDwhoRU**

### CODE REVIEW

File: plotmass/track.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                self.start_position = num
+
+        # Format arrays with single values
+        self.formatted_crv = numpy.append(self.get_crv_np()[self.get_start_position():self.get_array_length()], self.get_crv_np()[:self.get_end_position() + 1])
+        self.formatted_crv = abs(self.formatted_crv)
+
+        # Assign single values that require formatted arrays
+        crv_array_offset = numpy.append(self.get_formatted_crv()[self.get_start_position():self.get_array_length()], self.get_formatted_crv()[0:self.get_end_position() + 1])
+        abs_ofsts = [abs(x) for x in crv_array_offset]
+        offset_position = abs_ofsts.index(max(abs_ofsts))
```
Comment (CallumJHays) [inline plotmass/track.py:31]: Replace last 2 lines with `offset_position = np.argmax(abs(crv_array_offset))`

File: plotmass/track.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    def __init__(self, mat_file, crv_name, x_name):
+        self.track_data = loadmat(mat_file)
+        self.struct = list(dict.keys(self.track_data))[3]
+        self.crv_name = crv_name
+        self.x_name = x_name
+
+        # Read arrays into memory
+        self.crv_np = self.track_data[str(self.struct)][self.crv_name][0][0][0]
+        self.x = self.track_data[str(self.struct)][self.x_name][0][0][0]
```
Comment (CallumJHays) [inline plotmass/track.py:19]: general refactor option
```
extract_from_mat = lambda name: self.track_data[str(self.struct)][name][0][0][0]
self.crv_np = extract_from_mat(self.crv_name)
self.x = extract_from_mat(self.x_name)
```

File: plotmass/track.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+        # Single values and crv
+        self.crv = []
+        self.max_abs_value = max(abs(x) for x in self.crv_np)
+        for num, data in enumerate(self.crv_np):
+            self.crv.append(data)  
+            if abs(data) == self.get_max_abs_value():
+                self.start_position = num
```
Comment (CallumJHays) [inline plotmass/track.py]: As previously discussed
`self.start_position = np.argmax(abs(self.crv_np))`

File: plotmass/momentum.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+    def get_brake_a_lat(self):
+        return self.brake_a_lat
+
+    def get_velocity(self):
+        velocity = []
+
+        for i in range(len(self.accel_velocity)):
+            velocity.append(minimum(self.accel_velocity[i], self.brake_velocity[i]))
+        return velocity
```
Comment (CallumJHays) [inline plotmass/momentum.py:202]: one-liner
```
return [
    min(accel_vel, brake_vel)
    for accel_vel, brake_vel
    in zip(self.accel_velocity, self.brake_velocity)]
```

File: plotmass/momentum.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        crv = self.track.get_crv()
+
+        for i in range(len(self.accel_a_lat)):
+            a_lat.append(0)
+            if velocity[i] == self.accel_velocity[i]:
+                a_lat[i] = self.accel_a_lat[i]
+            else:
+                a_lat[i] = self.brake_a_lat[i]
+            
+            a_lat[i] = a_lat[i] * sign(crv[i])
```
Comment (CallumJHays) [inline plotmass/momentum.py:216]:
```
a_lat = [
    sign(crv_) * (accel if accel == vel else brake)
    for accel, vel, brake, crv_
    in zip(self.accel_a_lat, velocity, self.brake_a_lat, crv)]
```

Comment (CallumJHays) [reply plotmass/momentum.py:216]: or one-liner for the whole function:
```
return [
    sign(crv) * (accel if accel == vel else brake)
    for accel, vel, brake, crv
    in zip(self.accel_a_lat, self.get_velocity(), self.brake_a_lat, self.track.get_crv())]
```

File: plotmass/momentum.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    def get_a_long(self):
+        a_long = []
+        velocity = self.get_velocity()
+
+        for i in range(len(self.accel_a_long)):
+            a_long.append(0)
+            if velocity[i] == self.accel_velocity[i]:
+                a_long[i] = self.accel_a_long[i]
+            else:
+                a_long[i] = self.brake_a_long[i] * -1
```
Comment (CallumJHays) [inline plotmass/momentum.py:229]:
```
return [
    accel if accel == vel else brake + 1
    for accel, vel, brake
    in zip(self.accel_a_long, self.get_velocity(), self.brake_a_long)]
```

File: plotmass/track.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+from scipy.io import loadmat
+from scipy.signal import find_peaks
+import pandas as pd
+from math import *
+import numpy
```
Comment (CallumJHays) [inline plotmass/track.py:5]: industry convention:
`import numpy as np`

File: plotmass/track.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            self.crv.append(data)  
+            if abs(data) == self.get_max_abs_value():
+                self.start_position = num
+
+        # Format arrays with single values
+        self.formatted_crv = numpy.append(self.get_crv_np()[self.get_start_position():self.get_array_length()], self.get_crv_np()[:self.get_end_position() + 1])
+        self.formatted_crv = abs(self.formatted_crv)
+
+        # Assign single values that require formatted arrays
+        crv_array_offset = numpy.append(self.get_formatted_crv()[self.get_start_position():self.get_array_length()], self.get_formatted_crv()[0:self.get_end_position() + 1])
```
Comment (CallumJHays) [inline plotmass/track.py:29]:
```
reformat = lambda arr: np.append(arr[self.get_start_position():], arr[:self.get_start_position()])
self.formatted_crv = abs(reformat(self.get_crv_np()))
crv_array_offset = reformat(self.formatted_crv)
```

File: sim/views.py [hunk 9]
```diff
@@ -201,18 +230,26 @@ def speedcurvature(id, width=None, height=None):
...
-    graph_html, fastest_lap, min_speed, max_speed = plotMassSpeedCurvature(matfile, id.curvature, int(width), int(height), 9.81, id.mass, id.power, id.air_density, id.reference_area, id.coefficient_of_drag, id.coefficient_of_friction, id.coefficient_of_lift)
-    title = 'QUTMS | Speed v Curvature'
+    # Pickle graph & stats for download
+    simulation.pickle(simulation.plot_speed_curvature(), "graph_curvature.p")
 
+    # Update discord rich presence
     if rpc_activated:
-        RPC.update(state= str(int(id.mass)) + 'kg @ ' + str(int(id.power)) + 'W - ' + str(fastest_lap), details=str(id.name) + ' - Speed v Curvature', large_image="qut-logo")
-    return render_template('speedcurvature.html',min_speed=min_speed,max_speed=max_speed, graph_html=graph_html,title=title, name=id.name, fastest_lap=fastest_lap[2:], id=id)
+        RPC.update(state= str(int(id.mass)) + 'kg @ ' + str(int(id.power)) + 'W - ' + str(simulation.get_fastest_lap()), details=str(id.name) + ' - Speed v Curvature', large_image="qut-logo")
```
Comment (CallumJHays) [inline sim/views.py:251]: okay this is epic

File: plotmass/gg_diagram.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        self.fig = plt.figure()
+        self.ax = self.fig.add_subplot(1, 1, 1)
+        graph_size = (((window_h - 500) * (12 - 7)) / (1080 - 500)) + 7
+        self.fig.set_size_inches(graph_size, graph_size - 1)
+
+        # Title diagram and create legend
+        self.ax.set_title('G-G Diagram')
+        colour = self.gg_legend(a_lat, a_long)
+
+        bx, by, cx, cy, dx, dy, ex, ey, fx, fy, gx, gy, hx, hy, ix, iy, jx, jy = self.gg_colour_sorting(a_lat, a_long, colour)
```
Comment (CallumJHays) [inline plotmass/gg_diagram.py:19]: This is quite a mess but not an easy fix i can think of

File: plotmass/lateral_velocity.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        radius = []
+        for entry in self.crv_array:
+            radius.append(1/entry)       
+        top_speed = self.nthroot((2*P)/(p*Cd*A),3)
+        lat_velocity = []
+        for n in radius:
+            if 0.5*n*p*-Cl*A*mu > m:
+                vel = top_speed
+            else:
+                vel = sqrt((n*mu*m*g)/(m-0.5*n*p*-Cl*A*mu))
```
Comment (CallumJHays) [inline plotmass/lateral_velocity.py:37]: re-use calculation:
```
name_for_var = 0.5*n*p*-Cl*A*mu
if name_for_var > m:
    ...
else:
    vel = sqrt((n * mu * m * g) / (m - name_for_var))
```
```

File: plotmass/track.py (hunk 2 above)
Comment (LDwhoRU) [reply plotmass/track.py:19]: Keeping as is

File: plotmass/gg_diagram.py (hunk 10 above)
Comment (LDwhoRU) [reply plotmass/gg_diagram.py:19]: I'll eventually refactor into class attributes, this code is copied from the old design camp implementation from a new member

File: sim/views.py (hunk 9 above)
Comment (LDwhoRU) [reply sim/views.py:251]: :D

File: plotmass/track.py (hunk 8 above)
Comment (LDwhoRU) [reply plotmass/track.py:29]: I'll remove the getters but its fine as is

**Pull request merged**